File: StudentsMark/assignment 1.cs
using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Linq;

class Person
{
   public string FirstName { get; set; }
   public string Surname { get; set; }
   public float[] HomeworkScores { get; set; }
   public float ExamScore { get; set; }
   public float FinalGrade { get; set; }

   public Person(string firstName, string surname, float[] homeworkScores, float examScore)
   {
       FirstName = firstName;
       Surname = surname;
       HomeworkScores = homeworkScores;
       ExamScore = examScore;
       CalculateFinalGrade();
   }

   public Person(Person other)
   {
       FirstName = other.FirstName;
       Surname = other.Surname;
       HomeworkScores = (float[])other.HomeworkScores.Clone();
       ExamScore = other.ExamScore;
       FinalGrade = other.FinalGrade;
   }

   public Person Assign(Person other)
   {
       if (this != other)
       {
           FirstName = other.FirstName;
           Surname = other.Surname;
           HomeworkScores = (float[])other.HomeworkScores.Clone();
           ExamScore = other.ExamScore;
           FinalGrade = other.FinalGrade;
       }
       return this;
   }

   ~Person() { }

   public static Person InputData(string inputLine)
   {
       string[] row = Regex.Split(inputLine, @"\s+");
       string firstName = row[0];
       string surname = row[1];
       float[] homeworkScores = row.Skip(2).Take(row.Length - 2).Select(float.Parse).ToArray();
       float examScore = float.Parse(row[row.Length - 1]);

       return new Person(firstName, surname, homeworkScores, examScore);
   }

   public void OutputData()
   {
       Console.WriteLine($"{FirstName,-15} {Surname,-15} {FinalGrade,15:F2}");
   }

   public void CalculateFinalGrade()
   {
       float homeworkAverage = HomeworkScores.Average();
       FinalGrade = (0.4f * homeworkAverage) + (0.6f * ExamScore);
   }
}

class Program
{
   static void Main(string[] args)
   {
       string filePath = "students10000.txt";
       string[] lines = File.ReadAllLines(filePath);
       Console.WriteLine("{0,-15} {1,-15} {2,-15}", "Name", "Family", "Final_Point (Aver.)");

       for (int i = 1; i < lines.Length; i++)
       {
           Person student = Person.InputData(lines[i]);
           student.OutputData();
       }
   }
}

File: assignment 1.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <iomanip>

class Person
{
private:
    std::string FirstName;
    std::string Surname;
    std::vector<float> HomeworkScores;
    float ExamScore;
    float FinalGrade;

    void CalculateFinalGrade()
    {
        float homeworkAverage = std::accumulate(HomeworkScores.begin(), HomeworkScores.end(), 0.0f) / HomeworkScores.size();
        FinalGrade = 0.4f * homeworkAverage + 0.6f * ExamScore;
    }

public:
    Person(const std::string &firstName, const std::string &surname, const std::vector<float> &homeworkScores, float examScore)
        : FirstName(firstName), Surname(surname), HomeworkScores(homeworkScores), ExamScore(examScore)
    {
        CalculateFinalGrade();
    }

    Person(const Person &other)
        : FirstName(other.FirstName), Surname(other.Surname), HomeworkScores(other.HomeworkScores),
          ExamScore(other.ExamScore), FinalGrade(other.FinalGrade) {}

    Person &operator=(const Person &other)
    {
        if (this != &other)
        {
            FirstName = other.FirstName;
            Surname = other.Surname;
            HomeworkScores = other.HomeworkScores;
            ExamScore = other.ExamScore;
            FinalGrade = other.FinalGrade;
        }
        return *this;
    }

    ~Person() {}

    static Person InputData(const std::string &inputLine)
    {
        std::istringstream stream(inputLine);
        std::string firstName;
        std::string surname;
        std::vector<float> homeworkScores;
        float score;
        float examScore;

        stream >> firstName >> surname;
        while (stream >> score && stream.peek() != '\n')
        {
            homeworkScores.push_back(score);
        }
        examScore = homeworkScores.back();
        homeworkScores.pop_back();

        return Person(firstName, surname, homeworkScores, examScore);
    }

    void OutputData() const
    {
        std::cout << std::left << std::setw(15) << FirstName
                  << std::setw(15) << Surname
                  << std::right << std::setw(15) << std::fixed << std::setprecision(2) << FinalGrade << std::endl;
    }
};

int main()
{
    std::string filePath = "students10000.txt";
    std::ifstream file(filePath);
    std::string line;

    std::cout << std::left << std::setw(15) << "Name"
              << std::setw(15) << "Family"
              << std::setw(15) << "Final_Point (Aver.)" << std::endl;

    if (!file.is_open())
    {
        std::cerr << "Failed to open file." << std::endl;
        return 1;
    }

    getline(file, line); // skip first line for file title

    while (getline(file, line))
    {
        if (!line.empty())
        {
            Person student = Person::InputData(line);
            student.OutputData();
        }
    }

    return 0;
}
